cedar-lean/Cedar/Thm/Data/List.lean

```diff
@@ -755,21 +789,170 @@ theorem mapM_implies_nil {f : α → Except β γ} {as : List α}
+    simp only [List.mapM'] at h₁
+    cases h₂ : f hd <;> simp only [h₂, Except.bind_err, Except.bind_ok] at h₁
     cases h₃ : List.mapM' f tl <;>
     simp [h₃, pure, Except.pure] at h₁
 
 theorem mapM_head_tail {α β γ} {f : α → Except β γ} {x : α} {xs : List α} {y : γ} {ys : List γ} :
   (List.mapM f (x :: xs)) = Except.ok (y :: ys) →
   (List.mapM f xs) = Except.ok ys
 := by
-  simp [←List.mapM'_eq_mapM]
+  simp only [← mapM'_eq_mapM, mapM'_cons]
   cases h₁ : f x <;>
-  simp [h₁]
+  simp only [h₁, Except.bind_ok, Except.bind_err, false_implies]
   cases h₂ : mapM' f xs <;>
   simp [h₂, pure, Except.pure]
 
+theorem isSome_mapM {f : α → Option β} {xs : List α} :
+  Option.isSome (xs.mapM f) ↔ ∀ x ∈ xs, Option.isSome (f x)
+:= by
+  rw [← List.mapM'_eq_mapM]
+  constructor
+  case mp =>
+    induction xs
+    case nil => simp
+    case cons hd tl ih =>
+      simp only [mapM'_cons, Option.pure_def, Option.bind_eq_bind, mem_cons, forall_eq_or_imp]
+      intro h₁
+      unfold List.mapM' at h₁
+      constructor
+      case left =>
+        cases h₂ : (f hd)
+        <;> simp only [h₂, Option.pure_def, Option.bind_eq_bind,
+          Option.none_bind, Option.some_bind, Option.isSome_none] at h₁
+        simp
+      case right =>
+        intro a
+        apply ih ; clear ih
+        rw [Option.isSome_iff_exists] at h₁
+        replace ⟨as, h₁⟩ := h₁
+        cases h₂ : (f hd)
+        <;> simp only [h₂, Option.pure_def, Option.bind_eq_bind,
+          Option.some_bind, Option.none_bind, Option.bind_eq_some, Option.some.injEq] at h₁
+        replace ⟨bs, ⟨h₁, h₂⟩⟩ := h₁
+        simp only [Option.bind] at h₁
+        subst h₂
+        cases tl
+        case a.some.nil => simp
+        case a.some.cons hd tl =>
+          cases h₂ : f hd <;> simp [h₂] at *
+          cases h₂ : mapM' f tl <;> simp [h₂] at *
+  case mpr =>
+    intro h₁
+    induction xs
+    case nil => simp [List.mapM']
+    case cons hd tl ih =>
+      have h₂ := h₁ hd
+      simp only [mem_cons, true_or, forall_const] at h₂
+      rw [mapM'_cons]
+      cases h₃ : f hd <;> simp [h₃] at *
+      cases h₃ : mapM' f tl <;> simp [h₃] at *
+      case some => apply ih h₁
+
+/--
+  Corollary of `isSome_mapM`
+-/
+theorem mapM_eq_none {f : α → Option β} {xs : List α} :
+  xs.mapM f = none ↔ ∃ x ∈ xs, f x = none
+:= by
+  constructor
+  case mp =>
+    intro h₁
+    by_contra h₂
+    replace h₂ := forall_not_of_not_exists h₂
+    simp only [not_and] at h₂
+    rw [← Option.not_isSome_iff_eq_none] at h₁
+    rw [isSome_mapM] at h₁
+    apply h₁ ; clear h₁
+    intro v h₁
+    specialize h₂ v h₁
+    rw [← ne_eq] at h₂
+    rw [Option.ne_none_iff_isSome] at h₂
+    exact h₂
+  case mpr =>
+    intro h₁
+    replace ⟨v, h₁, h₂⟩ := h₁
+    rw [← Option.not_isSome_iff_eq_none]
+    intro h₃
+    rw [isSome_mapM] at h₃
+    specialize h₃ v h₁
+    simp [h₂] at h₃
+
+/--
+  Possibly the `mp` direction of `isSome_mapM` could be a corollary of this?
+-/
```

Comment:
For the corresponding theorems in Thm/Data/Map.lean, I did indeed prove the `mp` direction of `isSome_mapMOnValues` as a corollary of `mapMOnValues_eq_some`.  I held off on doing that refactoring here for the moment because (a) we do have a perfectly valid proof of `isSome_mapM` already, without `mapM_eq_some`, and (b) I wanted to wait until we completed the proof of `mapM_eq_some` before introducing more dependencies on it, just in case it turns out to be invalid or the statement need tweaking

Reply:
Since I've now resolved the problem with `mapM_eq_some`, I also refactored so `isSome_mapM.mp` is a corollary of it, which simplifies the proof substantially.